mlflow/models/container/__init__.py

```diff
@@ -86,22 +73,52 @@ def _serve():
+    has_env = False
+    if model_path:
+        model_config_path = os.path.join(model_path, "MLmodel")
+        model = Model.load(model_config_path)
+        # NOTE: this differs from _serve cause we always activate the env even if you're serving
+        # an mleap model
+        if pyfunc.FLAVOR_NAME not in model.flavors:
+            return
+        conf = model.flavors[pyfunc.FLAVOR_NAME]
+        if pyfunc.ENV in conf:
+            print("creating and activating custom environment")
+            env = conf[pyfunc.ENV]
+            env_path = os.path.join(model_path, env)
+            conda_create_model_env = "conda env create -n custom_env -f {}".format(env_path)
+            if Popen(["bash", "-c", conda_create_model_env]).wait() != 0:
+                raise Exception("Failed to create model environment.")
+            has_env = True
+    activate_cmd = ["source /miniconda/bin/activate custom_env"] if has_env else []
+    # NB: install gunicorn[gevent] from pip rather than from conda because gunicorn is already
+    # dependency of mlflow on pip and we expect mlflow to be part of the environment.
+    install_server_deps = ["pip install gunicorn[gevent]"]
+    if Popen(["bash", "-c", " && ".join(activate_cmd + install_server_deps)]).wait() != 0:
+        raise Exception("Failed to install serving dependencies into the model environment.")
+    if has_env and install_mlflow:
+        install_mlflow_cmd = [
+            "pip install /opt/mlflow/." if _container_includes_mlflow_source()
+            else "pip install mlflow=={}".format(MLFLOW_VERSION)
+        ]
+        if Popen(["bash", "-c", " && ".join(activate_cmd + install_mlflow_cmd)]).wait() != 0:
+            raise Exception("Failed to install mlflow into the model environment.")
+
+
 def _serve_pyfunc(model):
     conf = model.flavors[pyfunc.FLAVOR_NAME]
     bash_cmds = []
     if pyfunc.ENV in conf:
-        print("activating custom environment")
-        env = conf[pyfunc.ENV]
-        env_path_dst = os.path.join("/opt/mlflow/", env)
-        env_path_dst_dir = os.path.dirname(env_path_dst)
-        if not os.path.exists(env_path_dst_dir):
-            os.makedirs(env_path_dst_dir)
-        # TODO: should we test that the environment does not include any of the server dependencies?
-        # Those are gonna be reinstalled. should probably test this on the client side
-        shutil.copyfile(os.path.join(MODEL_PATH, env), env_path_dst)
-        os.system("conda env create -n custom_env -f {}".format(env_path_dst))
-        bash_cmds += ["source /miniconda/bin/activate custom_env"] + _server_dependencies_cmds()
-    nginx_conf = resource_filename(mlflow.sagemaker.__name__, "container/scoring_server/nginx.conf")
+        if not os.environ.get(DISABLE_ENV_CREATION) == "true":
+            _install_pyfunc_deps(MODEL_PATH, install_mlflow=True)
+        bash_cmds += ["source /miniconda/bin/activate custom_env"]
```

Comment:
Isn't this guaranteed to fail if `DISABLE_ENV_CREATION` is set and `pyfunc.ENV` is present in conf?

Reply:
DISABLE_ENV_CREATION is set iff you have already copied the model in and initialized its environment. E.g. if you build your image with mlflow models build-docker, the model is copied into the container and everything is set up. Sagemaker on the other hand mounts model at runtime so we need to create the environment then.

Reply:
O cool. This was another thing we ran into with trying to use the sagemaker code locally. We're restricted to a readonly file system in prod so we had to push the env setup into the image creation.